pkg/apis/scheduling/v1alpha1/types_location.go

```diff
@@ -0,0 +1,133 @@
+// +kubebuilder:printcolumn:name="Type",type=string,JSONPath=`.spec.type`,description="Type of the workspace"
+// +kubebuilder:printcolumn:name="Phase",type=string,JSONPath=`.status.phase`,description="The current phase (e.g. Scheduling, Initializing, Ready)"
+// +kubebuilder:printcolumn:name="URL",type=string,JSONPath=`.status.baseURL`,description="URL to access the workspace"
+type Location struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec LocationSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status LocationStatus `json:"status,omitempty"`
+}
+
+// LocationPoolReference defines a reference to a location pool.
+type LocationPoolReference struct {
+	// workspace is the absolute workspace name (e.g. root:company) of the pool this location belongs to.
+	Workspace string `json:"workspace"`
+	// name is the name of the location pool this location belongs to.
+	Name string `json:"name"`
+}
+
+// LocationSpec holds the desired state of the Location.
+type LocationSpec struct {
+	// pool is the LocationPool this location belongs to.
+	// +required
+	// +kubebuilder:Required
+	Pool LocationPoolReference `json:"pool"`
+
+	// type defines which class of objects this location can schedule. A typical
+	// type is "Workloads".
+	//
+	// +required
+	// +kubebuilder:Required
+	// +kubebuilder:validation:MinLength=1
+	Type string `json:"type,omitempty"`
+
+	PerLocationSpec `json:",inline"`
+}
+
+type PerLocationSpec struct {
+	// description is a human-readable description of the location.
+	//
+	// +optional
+	Description string `json:"description,omitempty"`
+
+	// labels is a set of labels that all instances in this location have in common.
+	//
+	// +optional
+	// +mapType=atomic
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// possibleNodeSelectorLabels is a list of labels that can be used to select an
+	// instance at this location in a placement object.
+	//
+	// +listType=map
+	// +listMapKey=key
+	PossibleNodeSelectorLabels []SchedulingProperty `json:"possibleNodeSelectorLabels,omitempty"`
+}
+
+// SchedulingProperty specified a label with key name and possible values.
+type SchedulingProperty struct {
+	// key is the name of the label.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	// +kubebuilder:Required
+	Key string `json:"key"`
+
+	// values are the possible values for this labels.
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	// +kubebuilder:Required
+	Values []LabelValues `json:"values"`
+
+	// description is a human readable description of the label.
+	//
+	// +optional
+	Description string `json:"description,omitempty"`
+}
+
+// LabelValues specifies a possible value for a property.
+type LabelValues string
+
+// LocationStatus defines the observed state of Location.
+type LocationStatus struct {
+	// instances is the number of actual instances at this location.
+	Instances uint32 `json:"instances,omitempty"`
```

Comment:
I am thinking a location always binds to a single instance?

Reply:
a namespace will bind to one instance of a location, but potentially multiple locations.